sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java

```diff
@@ -139,6 +146,11 @@ public void finalizeCheckpoint() throws IOException {
         // nothing to do
       }
     }
+
+    /* Get offset limit for unbounded source split checkpoint. */
+    default byte[] getOffsetLimit() {
```

Comment:
should we require this to be overridden (ie byte[0] is not valid) if offsetDeduplication is true?

Reply:
That sounds good. IIUC this means the default implementation would throw an exception and the runner (StreamingModeExecutionContext for Dataflow) must only call this method if the source enables offset deduplication. Updated PR to match this suggestion.

Reply:
Yup that's what I was thinking, thanks